### ISSUE TITLE
Program memory should not store JSON

### ISSUE THREAD

**Issue by adamchalmers** [KittyCAD/modeling-app#1130]
# Background

KCL program memory can store various different values, defined by the [MemoryItem enum](https://github.com/KittyCAD/modeling-app/blob/9f0ac5f6fddde509d7dbf1d44a2dca1949884068/src/wasm-lib/kcl/src/ast/types/literal_value.rs#L12). 

One case is "UserVal" which ultimately just stores `serde_json::Value` (see [definition of UserVal](https://github.com/KittyCAD/modeling-app/blob/455166459ab3ac2c9553c2945dc3b3564e1743c6/src/wasm-lib/kcl/src/executor.rs#L293)). User values are serialized into JSON before storing them in program memory. Reading from memory deserializes the value into its expected type, and returns a KCL type error if the value couldn't be deserialized into the right type.

# Problem

This works OK but leads to bugs when developing the modeling-app wasm-lib. Why? Because _so many_ types implement `serde::Deserialize` and you can try to deserialize a memory item into any of them. Even if those types never get written to memory ever! E.g. we never store a `SocketAddrV4` in KCL program memory, but the rust compiler won't warn you if you try to request that from KCL memory. But it should! It should be a rust-compile-time error to request something we'd never expect to find in KCL memory.

# Solution

We should remove the `UserVal` variant from `MemoryItem` and add new variants for KCL primitive types (numbers, strings, the KCL none type, etc).

This way, you ensure that reading values from memory requires that the value can at least be stored in memory.

**Comment by jtran** [comment]
I learned that in KCL, `0 / 0` is null! I learned this trying to make a test case that uses NaN or infinity. Couldn’t figure out those, but I got something else :smiling_imp: This happens because we represent all our numbers with a JSON type, and JSON doesn’t support NaN or infinity.

[Slack thread](https://kittycadworkspace.slack.com/archives/C06P9L5MCUD/p1722548755733469)

<img width="517" alt="Screenshot 2024-08-01 at 5 43 27 PM" src="https://github.com/user-attachments/assets/3d6a9ca9-b193-4f47-9b1e-2a8769decc6f">

Trying to use null as a number:

<img width="516" alt="Screenshot 2024-08-01 at 6 54 04 PM" src="https://github.com/user-attachments/assets/0aa68c9c-793f-4109-87c4-38e8fecf19bd">

**Comment by adamchalmers** [comment]
We basically want a model for what sorts of values exist in KCL. Here's my proposal.

* Integers
* Fractional numbers
* Strings
* Tag (ID and declaration)
* Closure (i.e. function + captured environment)
* Imported geometry
* KclArray (i.e. Vec<KclValue>)
* KclObject (i.e. HashMap<String, KclValue>)

If we do this, then KclValue's variants for SketchGroup, SketchGroupSet etc would be removed. Instead, SketchGroup would be serialized into KclObject. Functions which take a sketch group would have to check the KCL value being passed as an argument is a KclObject, then deserialize that object into a SketchGroup.

**Comment by jtran** [comment]
As a KCL user, I like the idea of making SketchGroup and other things be Objects like any other user-defined object. My only concern would be whether we can represent all the things we need internally. But if we can do that, I think this is a good idea.

**Comment by jtran** [comment]
I'm looking at the definition of `SketchGroup` et al now. The first thing that jumps out at me is that one of its fields is its `Path`s. A `Path` contains all the coordinates of the path, whether each segment is a line, tangential arc, angled line, etc. So this would need to be represented as KclObjects too, in addition to `Plane`, `Face`, and `ExtrudeGroup`.

If I were doing this, it seems big enough that I'd probably do it as a separate change, i.e. one PR to replace JSON with booleans, integers, floats, and strings. Then after it's merged, a second PR to collapse the rest into KclObjects.

Even just the first PR should fix #3338 and other issues.

**Comment by jtran** [comment]
To clarify, this is what I propose as a first step to reduce the downstream impact. I believe this would fix #3338 because the SketchGroup would be stored as a value in the `Object`'s HashMap, as a KclValue::SketchGroup (or possibly KclValue::SketchGroups plural). This avoids serialization to JSON. The serialization to JSON is problematic because nothing in executor knows how to deserialize again.

If you're doing the change, it's up to you whether you want to try to do it all at once, like in your comment above.

```rust
pub enum KclValue {
    // I don't want to get hung up on this.  I think we use null in a couple
    // places.  This is a replacement for that.  It would be nice to have
    // something for when a function has no meaningful return value, instead of
    // Option::None.
    Unit,
    Boolean(bool),
    String(String),
    Integer(i64),
    Float(f64),
    Array(Vec<KclValue>),
    Object(HashMap<String, KclValue>),

    // Everything below is unchanged.
    TagIdentifier(Box<TagIdentifier>),
    TagDeclarator(Box<TagDeclarator>),
    Plane(Box<Plane>),
    Face(Box<Face>),
    SketchGroup(Box<SketchGroup>),
    SketchGroups {
        value: Vec<Box<SketchGroup>>,
    },
    ExtrudeGroup(Box<ExtrudeGroup>),
    ExtrudeGroups {
        value: Vec<Box<ExtrudeGroup>>,
    },
    ImportedGeometry(ImportedGeometry),
    #[ts(skip)]
    Function {
        #[serde(skip)]
        func: Option<MemoryFunction>,
        expression: Box<FunctionExpression>,
        memory: Box<ProgramMemory>,
        #[serde(rename = "__meta")]
        meta: Vec<Metadata>,
    },
}
```

**Comment by adamchalmers** [comment]
The experiment of "serialize all values into JSON and just store JSON everywhere" is not looking good. Lee's L-system program is very slow because it spends most of its time repeatedly deserializing sketchgroups. I also don't think this approach will let us store KCL functions -- how would you serialize a KCL function into JSON?

So, I think we should return to the proposal above, i.e. a `KclValue` enum with no `UserValue` variant. We'd replace it with variants for each primitive type (e.g. numeric types, string, bool) and collections (i.e. object and array).
